### PR TITLE
support redirection url with bookmark and query

### PR DESCRIPTION
**Pull request by 928PJY**
- support redirection url with bookmark and query
- add hint for the following case.
  1. Redirection URL is `/basepath/a/` and there is no file published to `/basepath/a/`, but there a file publish to `/basepath/a`(DHS will fallback from  `/basepath/a/` to  `/basepath/a`)
  2. Redirection URL is `/basepath/a` and there is no file published to `/basepath/a`, but there a file publish to `/basepath/a/`(DHS will fallback from  `/basepath/a` to  `/basepath/a/`)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/5047)

### CODE REVIEW

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 1]
```diff
@@ -109,36 +110,51 @@ void AddRedirections(Dictionary<string, SourceInfo<string>> items, bool redirect
...
-            foreach (var (redirectionUrl, redirect) in redirectionsWithDocumentId)
+                .Concat(redirections.Select(item => item.redirection.SiteUrl)).ToHashSet();
+            foreach (var (originalRedirectUrl, normalizedRedirectUrl) in redirectionsWithDocumentId)
             {
-                if (!publishUrls.Contains(redirect.RedirectionUrl))
+                if (!publishUrls.Contains(normalizedRedirectUrl))
                 {
-                    errorLog.Write(Errors.RedirectionUrlNotExisted(redirectionUrl));
+                    string candidateRedirectionUrl = null;
+                    if (normalizedRedirectUrl.EndsWith("/") && publishUrls.Contains(normalizedRedirectUrl.Substring(0, normalizedRedirectUrl.Length - 1)))
```
Comment (live1206) [inline src/docfx/build/redirection/RedirectionMap.cs]: no need to check `EndsWith`?

Comment (928PJY) [reply src/docfx/build/redirection/RedirectionMap.cs]: why, what `/ab` and `/a`

**Comment by 928PJY** [comment]
https://github.com/dotnet/docfx/issues/5048

File: src/docfx/build/redirection/RedirectionMap.cs [hunk 2]
```diff
@@ -120,25 +120,31 @@ void AddRedirections(Dictionary<string, SourceInfo<string>> items, bool redirect
             }
         }
 
+        private static string NormalizeRedirectUrl(string redirectionUrl)
+        {
+            var (url, _, _) = UrlUtility.SplitUrl(redirectionUrl);
+            return url.EndsWith("/index", PathUtility.PathComparison) ? url.Substring(0, url.Length - "index".Length) : url;
```
Comment (yufeih) [inline src/docfx/build/redirection/RedirectionMap.cs]: combine bookmark query after `url.Substring`?

File: docs/specs/metadata.yml [hunk 3]
```diff
@@ -114,19 +114,33 @@ outputs:
 inputs:
   docfx.yml: |
     documentId:
-      sourceBasePath: sccm
-    name: sccm
```
Comment (OsmondJiang) [inline docs/specs/metadata.yml]: don''t change this test, it's from real world, for specific purpose. Add a new test?

File: src/docfx/Errors.cs [hunk 4]
```diff
@@ -43,7 +43,11 @@ public static Error RedirectionUrlConflict(SourceInfo<string> source)
         /// The dest to redirection url does not match any files's publish URL, but the redirect_with_id flag has been set as true
         /// </summary>
         public static Error RedirectionUrlNotExisted(SourceInfo<string> source)
-            => new Error(ErrorLevel.Suggestion, "redirection-url-not-existed", $"The redirect url '{source}' does not match any file's publish URL, but the redirect_with_id flag has been set as true", source);
+        {
+            var msg = $"The redirect url '{source}' does not match any file's publish URL, but the redirect_with_id flag has been set as true,"
```
Comment (OsmondJiang) [inline src/docfx/Errors.cs]: we don;t need that any more

File: docs/specs/metadata.yml (hunk 3 above)
Comment (928PJY) [reply docs/specs/metadata.yml]: from the title, this test case is for the same purpose.
and spec test doesn't require real world case, right? more simple, more readable

---
In reply to: [320657688](https://github.com/dotnet/docfx/pull/5047#discussion_r320657688) [](ancestors = 320657688)

Comment (OsmondJiang) [reply docs/specs/metadata.yml]: OK, then how do you know the document id of X is really from A?

---
In reply to: [321588571](https://github.com/dotnet/docfx/pull/5047#discussion_r321588571) [](ancestors = 321588571,320657688)

Comment (928PJY) [reply docs/specs/metadata.yml]: manually verified😆

**Pull request merged**